Library/Backends/ZendEngine2/Backend.php

```diff
@@ -253,17 +264,39 @@ public function onPostCompile($method, CompilationContext $context)
         $codePrinter = $context->codePrinter;
         if (preg_match('/^zephir_init_properties/', $method->getName())) {
             $codePrinter->increaseLevel();
-            $codePrinter->output('return Z_OBJVAL_P(this_ptr);');
+            $codePrinter->output('return retval;');
             $codePrinter->decreaseLevel();
             $codePrinter->output('}');
             $codePrinter->decreaseLevel();
         }
+        else {
+            $cleanup = '';
+            foreach ($context->symbolTable->getVariables() as $variable) {
+                if ($variable->getName() != 'this_ptr' && $variable->getName() != 'return_value' && $variable->getName() != 'return_value_ptr') {
```

Comment:
The current implementation is much like a hack :-(

There is no `cleanup()` method or something like this, therefore we have to (ab)use `onPostCompile()`. But the issue is that in `postCompile()` the entire method has already been generated, and we have to deal with the code in the text form.

The current logic is as follows:
* we get the list of all used variables and filter only the local ones;
* for every local variable we add a call to `zval_dtor()`;
* all these calls are saved into a string;
* then we get the text code from the code printer and insert out string with `zval_dtor()`s right before the `RETURN_xxx` statement.

I know that this is not an elegant solution, but with the current implementation of the compiler I was unable to come up with something more beautiful :-(